### PR TITLE
ensure tasks have been resolved before run task commands are executed, only show default build tasks in quick-pick

### LINKED ISSUE #173384: Build task not found sometimes
* Install Java extension pack
* Open VS Code repo
* Cmd+Shift+B
* 🐛 Java extension prompts error and the task detection never finishes

### LINKED ISSUE #179684: VS Code won`t show tasks defined in the .code-workspace file
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 
  Version: 1.77.1 (user setup)
  Commit: b7886d7461186a5eac768481578c1d7ca80e2d21
  Date: 2023-04-04T23:21:11.906Z
  Electron: 19.1.11
  Chromium: 102.0.5005.196
  Node.js: 16.14.2
  V8: 10.2.154.26-electron.0
  OS: Windows_NT x64 10.0.19044
  Sandboxed: No

- OS Version: Windows 10 Enterprise | Version 21H2 | Build 19044.2728

Steps to Reproduce:

1.  Create a .code-workspace file and open the workspace (mine is a multi-root workspace).
2.  Create a custom task under "tasks" property (example below) image
4.  Open the command palette and type "Tasks: Run Task"

Ideally you should see your workspace level created tasks under the command palette, but it only shows other options (image below):

image

One thing that I noticed is that if you create an empty tasks.json file in any of the projects inside the workspace it will show the tasks. It also will show tasks if you delete an empty tasks.json file. But after closing vs code it will still not be showing the tasks under command palette.

image

### PR DESCRIPTION
**Pull request by meganrogge**
fix #173384
fix #179684

see lengthy investigation / explanation starting with https://github.com/microsoft/vscode/issues/173384#issuecomment-1604766267

### CODE REVIEW

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 1]
```diff
@@ -335,6 +335,7 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 		this._waitForSupportedExecutions = new Promise(resolve => {
 			once(this._onDidRegisterSupportedExecutions.event)(() => resolve());
 		});
+
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:338]:
```suggestion
```

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Comment by Tyriar** [comment]
A more detailed PR title would be helpful as we might need to reference this later

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 2]
```diff
@@ -2875,24 +2875,21 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
...
 		const defaults: Task[] = [];
 		for (const task of tasks) {
 			// At this point (assuming taskGlobsInList is true) there are tasks with matching globs, so only put those in defaults
 			if (taskGlobsInList && typeof (task.configurationProperties.group as TaskGroup).isDefault === 'string') {
 				defaults.push(task);
 			} else if (!taskGlobsInList && (task.configurationProperties.group as TaskGroup).isDefault === true) {
 				defaults.push(task);
-			} else {
-				none.push(task);
 			}
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2889]: Yes, before it was adding a bunch of random tasks to the list when the build task would be run if it wasn't found in the first pass (IE the `group` property hadn't been resolved or in the case of a multi-root repo).

IMO, it was just a misunderstanding of what this command should do.

[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:2889]: https://github.com/microsoft/vscode/issues/88106

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 3]
```diff
@@ -344,7 +344,7 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 				}
 			}));
 		}
-
+		this._registerCommands().then(() => TaskCommandsRegistered.bindTo(this._contextKeyService).set(true));
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts]: that was an old change I made. reverted

**Review by Tyriar — APPROVED** [review]

**Pull request merged**